### ISSUE TITLE
Failed to apply update 43: Failed adding unique index to storage_pools_config and networks_config tables

### ISSUE THREAD

**Issue by alexejk** [lxc/lxd#8807]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04 (focal) LTS
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.0-72-generic
   * LXC version: 4.14
   * LXD version: 4.14
   * Storage backend in use: btrfs

# Issue description

When system was restarted, LXD fails to start with a timeout.
Running `lxd --debug --group lxd` produces following output:

```
INFO[05-24|16:15:22] LXD 4.14 is starting in normal mode      path=/var/snap/lxd/common/lxd
INFO[05-24|16:15:22] Kernel uid/gid map: 
INFO[05-24|16:15:22]  - u 0 0 4294967295 
INFO[05-24|16:15:22]  - g 0 0 4294967295 
INFO[05-24|16:15:22] Configured LXD uid/gid map: 
INFO[05-24|16:15:22]  - u 0 1000000 1000000000 
INFO[05-24|16:15:22]  - g 0 1000000 1000000000 
INFO[05-24|16:15:22] Kernel features: 
INFO[05-24|16:15:22]  - closing multiple file descriptors efficiently: no 
INFO[05-24|16:15:22]  - netnsid-based network retrieval: yes 
INFO[05-24|16:15:22]  - pidfds: yes 
INFO[05-24|16:15:22]  - uevent injection: yes 
INFO[05-24|16:15:22]  - seccomp listener: yes 
INFO[05-24|16:15:22]  - seccomp listener continue syscalls: yes 
INFO[05-24|16:15:22]  - seccomp listener add file descriptors: no 
INFO[05-24|16:15:22]  - attach to namespaces via pidfds: no 
INFO[05-24|16:15:22]  - safe native terminal allocation : yes 
INFO[05-24|16:15:22]  - unprivileged file capabilities: yes 
INFO[05-24|16:15:22]  - cgroup layout: hybrid 
WARN[05-24|16:15:22]  - Couldn't find the CGroup blkio.weight, disk priority will be ignored 
WARN[05-24|16:15:22]  - Couldn't find the CGroup memory swap accounting, swap limits will be ignored 
INFO[05-24|16:15:22]  - shiftfs support: yes 
INFO[05-24|16:15:22] Initializing local database 
DBUG[05-24|16:15:22] Refreshing local trusted certificate cache 
INFO[05-24|16:15:22] Set client certificate to server certificate ea45d0496dfdab05222567fae41fa2ab9d243fc34f75b52df35b6fb2fbc99407 
DBUG[05-24|16:15:22] Initializing database gateway 
DBUG[05-24|16:15:22] Start database node                      id=1 address= role=voter
WARN[05-24|16:15:23] Failed settting up shared mounts: no such file or directory 
INFO[05-24|16:15:23] Starting /dev/lxd handler: 
INFO[05-24|16:15:23]  - binding devlxd socket                 socket=/var/snap/lxd/common/lxd/devlxd/sock
INFO[05-24|16:15:23] REST API daemon: 
INFO[05-24|16:15:23]  - binding Unix socket                   socket=/var/snap/lxd/common/lxd/unix.socket
INFO[05-24|16:15:23]  - binding TCP socket                    socket=[::]:8443
INFO[05-24|16:15:23] Initializing global database 
DBUG[05-24|16:15:23] Dqlite: attempt 0: server 1: connected 
INFO[05-24|16:15:23] Updating the LXD global schema. Backup made as "global.bak" 
DBUG[05-24|16:15:23] Updating global DB schema from 30 to 31 
DBUG[05-24|16:15:23] Updating global DB schema from 31 to 32 
DBUG[05-24|16:15:23] Updating global DB schema from 32 to 33 
DBUG[05-24|16:15:23] Updating global DB schema from 33 to 34 
DBUG[05-24|16:15:23] Updating global DB schema from 34 to 35 
DBUG[05-24|16:15:23] Updating global DB schema from 35 to 36 
DBUG[05-24|16:15:23] Updating global DB schema from 36 to 37 
DBUG[05-24|16:15:23] Updating global DB schema from 37 to 38 
DBUG[05-24|16:15:23] Updating global DB schema from 38 to 39 
DBUG[05-24|16:15:23] Updating global DB schema from 39 to 40 
DBUG[05-24|16:15:23] Updating global DB schema from 40 to 41 
DBUG[05-24|16:15:23] Updating global DB schema from 41 to 42 
WARN[05-24|16:15:23] Found duplicated network config rows     rowCount=2 dupeRowIDs=21,27 networkID=1 nodeID=-1 key=ipv4.address value=10.0.3.1/24
WARN[05-24|16:15:23] Deleted duplicated network config row    rowID=21 networkID=1 nodeID=-1 key=ipv4.address value=10.0.3.1/24 rowCount=2
WARN[05-24|16:15:23] Found duplicated network config rows     key=ipv4.address value=10.0.4.1/24 rowCount=5 dupeRowIDs=7,9,11,14,17 networkID=1 nodeID=-1
WARN[05-24|16:15:23] Deleted duplicated network config row    nodeID=-1 key=ipv4.address value=10.0.4.1/24 rowCount=5 rowID=7 networkID=1
WARN[05-24|16:15:23] Deleted duplicated network config row    networkID=1 nodeID=-1 key=ipv4.address value=10.0.4.1/24 rowCount=5 rowID=9
WARN[05-24|16:15:23] Deleted duplicated network config row    rowCount=5 rowID=11 networkID=1 nodeID=-1 key=ipv4.address value=10.0.4.1/24
WARN[05-24|16:15:23] Deleted duplicated network config row    networkID=1 nodeID=-1 key=ipv4.address value=10.0.4.1/24 rowCount=5 rowID=14
WARN[05-24|16:15:23] Found duplicated network config rows     networkID=1 nodeID=-1 key=ipv4.nat value=true rowCount=7 dupeRowIDs=8,10,12,15,18,22,28
WARN[05-24|16:15:23] Deleted duplicated network config row    rowCount=7 rowID=8 networkID=1 nodeID=-1 key=ipv4.nat value=true
WARN[05-24|16:15:23] Deleted duplicated network config row    networkID=1 nodeID=-1 key=ipv4.nat value=true rowCount=7 rowID=10
WARN[05-24|16:15:23] Deleted duplicated network config row    value=true rowCount=7 rowID=12 networkID=1 nodeID=-1 key=ipv4.nat
WARN[05-24|16:15:23] Deleted duplicated network config row    value=true rowCount=7 rowID=15 networkID=1 nodeID=-1 key=ipv4.nat
WARN[05-24|16:15:23] Deleted duplicated network config row    networkID=1 nodeID=-1 key=ipv4.nat value=true rowCount=7 rowID=18
WARN[05-24|16:15:23] Deleted duplicated network config row    networkID=1 nodeID=-1 key=ipv4.nat value=true rowCount=7 rowID=22
WARN[05-24|16:15:23] Found duplicated network config rows     rowCount=5 dupeRowIDs=13,16,19,23,25 networkID=1 nodeID=-1 key=ipv6.address value=none
WARN[05-24|16:15:23] Deleted duplicated network config row    value=none rowCount=5 rowID=13 networkID=1 nodeID=-1 key=ipv6.address
WARN[05-24|16:15:23] Deleted duplicated network config row    networkID=1 nodeID=-1 key=ipv6.address value=none rowCount=5 rowID=16
WARN[05-24|16:15:23] Deleted duplicated network config row    key=ipv6.address value=none rowCount=5 rowID=19 networkID=1 nodeID=-1
WARN[05-24|16:15:23] Deleted duplicated network config row    rowCount=5 rowID=23 networkID=1 nodeID=-1 key=ipv6.address value=none
WARN[05-24|16:15:23] Found duplicated network config rows     networkID=1 nodeID=-1 key=ipv6.nat value=false rowCount=2 dupeRowIDs=6,26
WARN[05-24|16:15:23] Deleted duplicated network config row    rowCount=2 rowID=6 networkID=1 nodeID=-1 key=ipv6.nat value=false
WARN[05-24|16:15:23] Found duplicated network config rows     value=true rowCount=2 dupeRowIDs=20,24 networkID=1 nodeID=-1 key=ipv6.nat
WARN[05-24|16:15:23] Deleted duplicated network config row    rowCount=2 rowID=20 networkID=1 nodeID=-1 key=ipv6.nat value=true
DBUG[05-24|16:15:23] Updating global DB schema from 42 to 43 
DBUG[05-24|16:15:23] Updating global DB schema from 43 to 44 
DBUG[05-24|16:15:23] Database error: &errors.errorString{s:"failed to apply update 43: Failed adding unique index to storage_pools_config and networks_config tables: UNIQUE constraint failed: index 'networks_unique_network_id_node_id_key'"} 
EROR[05-24|16:15:23] Failed to start the daemon: failed to open cluster database: failed to ensure schema: failed to apply update 43: Failed adding unique index to storage_pools_config and networks_config tables: UNIQUE constraint failed: index 'networks_unique_network_id_node_id_key' 
INFO[05-24|16:15:23] Starting shutdown sequence 
INFO[05-24|16:15:23] Stop database gateway 
INFO[05-24|16:15:23] Stopping REST API handler: 
INFO[05-24|16:15:23]  - closing socket                        socket=[::]:8443
INFO[05-24|16:15:23]  - closing socket                        socket=/var/snap/lxd/common/lxd/unix.socket
INFO[05-24|16:15:23] Stopping /dev/lxd handler: 
INFO[05-24|16:15:23]  - closing socket                        socket=/var/snap/lxd/common/lxd/devlxd/sock
DBUG[05-24|16:15:23] Not unmounting temporary filesystems (containers are still running) 
Error: failed to open cluster database: failed to ensure schema: failed to apply update 43: Failed adding unique index to storage_pools_config and networks_config tables: UNIQUE constraint failed: index 'networks_unique_network_id_node_id_key'
```

**Comment by stgraber** [comment]
Could you install sqlite3 and show the output of:
 - sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM networks;"
 - sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM networks_nodes;"

**Comment by stgraber** [comment]
It looks like your database has some kind of duplicate entry that's breaking that schema update. Once we figure out what it is, we can give you a small SQL patch that LXD can load on startup and unblock the update.

**Comment by tomponline** [comment]
This seems like basically a duplicate of https://github.com/lxc/lxd/issues/8260 although the error is different because it is occurring after the DB schema patch to add the unique index (to prevent it from occurring again) has been added.

As @stgraber said, we need to see the contents of those tables now so we can prepare a custom DB patch you can apply to fix the issue.

**Comment by stgraber** [comment]
@alexejk ping

**Comment by hi-tech** [comment]
Having the similar (or same) issue - both sqlite commands yield the same syntax error in my case:

Full error
```
Failed to start the daemon: failed to open cluster database: failed to ensure schema: failed to apply update 43: Failed adding unique index to storage_pools_config and networks_config tables: UNIQUE constraint failed: index 'networks_unique_network_id_node_id_key' 
Error: failed to open cluster database: failed to ensure schema: failed to apply update 43: Failed adding unique index to storage_pools_config and networks_config tables: UNIQUE constraint failed: index 'networks_unique_network_id_node_id_key'

```

```
sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM networks;"
Error: malformed database schema (projects_used_by_ref) - near "(": syntax error
```

```
sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM networks_nodes;"
Error: malformed database schema (projects_used_by_ref) - near "(": syntax error
```

```
sqlite> .dump projects_used_by_ref
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE VIEW projects_used_by_ref (name,
    value) AS
  SELECT projects.name,
    printf('/1.0/instances/%s?project=%s',
    "instances".name,
    projects.name)
    FROM "instances" JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/images/%s?project=%s',
    images.fingerprint,
    projects.name)
    FROM images JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/storage-pools/%s/volumes/custom/%s?project=%s&target=%s',
    storage_pools.name,
    storage_volumes.name,
    projects.name,
    nodes.name)
    FROM storage_volumes JOIN storage_pools ON storage_pool_id=storage_pools.id JOIN nodes ON node_id=nodes.id JOIN projects ON project_id=projects.id WHERE storage_volumes.type=2 UNION
  SELECT projects.name,
    printf('/1.0/profiles/%s?project=%s',
    profiles.name,
    projects.name)
    FROM profiles JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/networks/%s?project=%s',
    networks.name,
    projects.name)
    FROM networks JOIN projects ON project_id=projects.id;
COMMIT;

```

```
sqlite> .schema projects_used_by_ref
CREATE VIEW projects_used_by_ref (name,
    value) AS
  SELECT projects.name,
    printf('/1.0/instances/%s?project=%s',
    "instances".name,
    projects.name)
    FROM "instances" JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/images/%s?project=%s',
    images.fingerprint,
    projects.name)
    FROM images JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/storage-pools/%s/volumes/custom/%s?project=%s&target=%s',
    storage_pools.name,
    storage_volumes.name,
    projects.name,
    nodes.name)
    FROM storage_volumes JOIN storage_pools ON storage_pool_id=storage_pools.id JOIN nodes ON node_id=nodes.id JOIN projects ON project_id=projects.id WHERE storage_volumes.type=2 UNION
  SELECT projects.name,
    printf('/1.0/profiles/%s?project=%s',
    profiles.name,
    projects.name)
    FROM profiles JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/networks/%s?project=%s',
    networks.name,
    projects.name)
    FROM networks JOIN projects ON project_id=projects.id;
sqlite> 
```

**Comment by stgraber** [comment]
@hi-tech can you send me a tarball of `/var/snap/lxd/common/lxd/database` to stgraber at ubuntu dot com?

**Comment by alexejk** [comment]
Sorry for delay. I tried getting DB dump but given lxd doesn't start i can't use `lxd sql` to access it. Any guidance on how to perform the dump in this situation?
Thanks :)

**Comment by tomponline** [comment]
@alexejk the commands to run are here https://github.com/lxc/lxd/issues/8807#issuecomment-847462675

**Comment by alexejk** [comment]
I'm blind, my apologies.

> `sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM networks;"`

```
1|lxdbr0||1
```

> `sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM networks_nodes;"`

```
1|1|1
```

**Comment by tomponline** [comment]
@alexejk no worries :)

Also:

```
sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin 'select * from networks_config'
```

**Comment by alexejk** [comment]
> `sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin 'select * from networks_config;'`

```
5|1||ipv6.address|fd42:e9b:4521:6952::1/64
6|1||ipv6.nat|false
7|1||ipv4.address|10.0.4.1/24
8|1||ipv4.nat|true
9|1||ipv4.address|10.0.4.1/24
10|1||ipv4.nat|true
11|1||ipv4.address|10.0.4.1/24
12|1||ipv4.nat|true
13|1||ipv6.address|none
14|1||ipv4.address|10.0.4.1/24
15|1||ipv4.nat|true
16|1||ipv6.address|none
17|1||ipv4.address|10.0.4.1/24
18|1||ipv4.nat|true
19|1||ipv6.address|none
20|1||ipv6.nat|true
21|1||ipv4.address|10.0.3.1/24
22|1||ipv4.nat|true
23|1||ipv6.address|none
24|1||ipv6.nat|true
25|1||ipv6.address|none
26|1||ipv6.nat|false
27|1||ipv4.address|10.0.3.1/24
28|1||ipv4.nat|true
```

**Comment by tomponline** [comment]
@alexejk ok so you can see there are a lot of duplicate fields (the 3rd and 4th columns) for network ID 1 (the 2nd column).

There should only be a maximum of 1 config value per field (e.g. `ipv4.address = 10.0.4.1/24`) and they certainly can't be different (that is the error you're getting because we cannot apply the unique index).

So...

You need to decide on which set of config rows represents your desired config (i.e some rows say `ipv4.nat=true` and others say `ipv4.nat=true`, also some say `ipv4.address=10.0.4.1/24` but others say `ipv4.address=10.0.3.1/24`, and some say `ipv6.address=none` and others say `ipv6.address=fd42:e9b:4521:6952::1/64`).

Once you've done that we can craft a patch file that will selectively remove the offending row IDs and leave only the desired ones.

**Comment by alexejk** [comment]
Interesting that it ended up in that state somehow. Quite a while ago I was creating a custom a new network and assigned to some containers but that was about it...

The primary network for me is `ipv4.address=10.0.3.1/24` (and I think `ipv4.nat=true`). I assume once it starts I can adjust if the NAT value is wrong.

**Comment by tomponline** [comment]
@alexejk there was quite an old bug that allowed duplicates to get into the DB, this was fixed some time ago (see the other issue I linked to earlier), but unfortunately we couldn't automatically fix conflicting values.

Here's a post from the forums showing how to craft the patch file:
https://discuss.linuxcontainers.org/t/failed-to-load-network-lxdbr0-duplicate-config-row-found-for-key-dns-mode-for-network-id-1/10101/5

**Comment by tomponline** [comment]
So the rows you most likely want to keep are:

```
25|1||ipv6.address|none
26|1||ipv6.nat|false
27|1||ipv4.address|10.0.3.1/24
28|1||ipv4.nat|true
```

**Comment by alexejk** [comment]
Thank you! Adding the patch statement got LXD started so now I can fix any things that are wrong (if any). Appreciate the help!

**Comment by hi-tech** [comment]
> @hi-tech can you send me a tarball of `/var/snap/lxd/common/lxd/database` to stgraber at ubuntu dot com?

Thanks, you can ignore - I had the same duplicate network issue (results of cloning hosts) and the patch above worked.

However, It seems "column-name list" in sqlite views was [added in sqlite 3.9.x](https://www.sqlite.org/lang_createview.html)
Some of our LXDs happening to be running on Centos 7.8/9 hosts, with an ancient sqlite 3.7.
Backing up / compiling / sym-linking newest sqlite 3.35 fixed the 2nd issue. We are migrating away from Centos hosts in any event.

**Comment by stgraber** [comment]
Ah, okay, a different version of sqlite in the snap and outside of it would indeed explain the confusing output you got.

Happy you got it sorted out!

**Comment by hi-tech** [comment]
Ahh, didn't realize snap has its own version of sqlite. Just put back the original sqlite3 binary and all is still well. 

Thanks again for that quick sql patch

**Comment by stgraber** [comment]
Yep, that's the main thing we like with the snap package. Doesn't matter what distro you run it on, you're running the exact same binaries and if you tell us what revision on what architecture, we can replicate your environment almost perfectly.

Obviously the kernel you're using may differ and that's the main source of issues when it comes to this package. But having the majority of our users on the exact same sqlite3, qemu, lxc, dnsmasq, ... sure is making things easier :)
